metal/resource_metal_port.go

```diff
@@ -0,0 +1,342 @@
+			return err
+		}
+	}
+
+	// remove vlans
+	vlansToRemove := difference(
+		attachedVlanIds(port),
+		specifiedVlanIds(d),
+	)
+	err = processVlansOnPort(port.ID, vlansToRemove, client.DevicePorts.Unassign)
+	if err != nil {
+		return err
+	}
+
+	// disbond
+	bonded, bondedOk := d.GetOk("bonded")
+	if bondedOk && !bonded.(bool) && port.Data.Bonded {
+		port, _, err = client.DevicePorts.Disbond(port, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	// convert to layer2
+	isLayer2 := contains(l2Types, port.NetworkType)
+
+	if l2Ok && l2.(bool) && !isLayer2 {
+		port, _, err = client.Ports.ConvertToLayerTwo(port.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// convert to layer3
+	if l2Ok && !l2.(bool) && isLayer2 {
+		ips := []packngo.AddressRequest{
+			{AddressFamily: 4, Public: true},
+			{AddressFamily: 4, Public: false},
+			{AddressFamily: 6, Public: true},
+		}
+		port, _, err = client.Ports.ConvertToLayerThree(port.ID, ips)
+		if err != nil {
+			return err
+		}
+	}
+
+	// constructive phase - bond creation and vlan assignment
+	// ======================================================
+
+	port, err = getPortByResourceData(d, client)
+	if err != nil {
+		return err
+	}
+
+	// bond
+	if bondedOk && bonded.(bool) && !port.Data.Bonded {
+		port, _, err = client.DevicePorts.Bond(port, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	// assign VLANs
+	vlansToAssign := difference(
+		specifiedVlanIds(d),
+		attachedVlanIds(port),
+	)
+	err = processVlansOnPort(port.ID, vlansToAssign, client.DevicePorts.Assign)
+	if err != nil {
+		return err
+	}
+
+	// assign Native VLAN
+	currentNative = getCurrentNative(port)
+	specifiedNative = getSpecifiedNative(d)
+	if (currentNative != specifiedNative) && currentNative != "" {
+		par := packngo.PortAssignRequest{
+			PortID:           port.ID,
+			VirtualNetworkID: specifiedNative,
+		}
+		_, _, err = client.DevicePorts.AssignNative(&par)
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceMetalPortRead(d, meta)
+}
+
+func resourceMetalPortRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*packngo.Client)
+	port, err := getPortByResourceData(d, client)
+	if err != nil {
+		return err
+	}
+	m := map[string]interface{}{
+		"layer2":            contains(l2Types, port.NetworkType),
+		"type":              port.Type,
+		"name":              port.Name,
+		"network_type":      port.NetworkType,
+		"mac":               port.Data.MAC,
+		"bonded":            port.Data.Bonded,
+		"disbond_supported": port.DisbondOperationSupported,
+	}
+
+	if port.NativeVirtualNetwork != nil {
+		m["native_vlan_id"] = port.NativeVirtualNetwork.ID
+	}
+
+	if len(port.AttachedVirtualNetworks) > 0 {
+		vlans := []string{}
+		for _, n := range port.AttachedVirtualNetworks {
+			vlans = append(vlans, n.ID)
+		}
+		m["vlan_ids"] = vlans
+	}
+
+	if port.Bond != nil {
+		m["bond_id"] = port.Bond.ID
+		m["bond_name"] = port.Bond.Name
+	}
+
+	d.SetId(port.ID)
+	return setMap(d, m)
```

Comment:
The `setMap` helper takes functions where the `if` blocks could be resolved.  It's the pattern that I've been using. 

I don't see anything wrong with the approach you've taken. It's flatter.